### PR TITLE
Refactor stbt-run

### PR DESCRIPTION
**Pull request by wmanley**
This clarifies the different responsibilities of stbt-run and the data flow.

It does remove `init_run`.  I don't know if we consider this to be public API?

Another user facing change is that it says "Saved screenshot" before the traceback rather than after.

### CODE REVIEW

File: stbt-run [hunk 1]
```diff
@@ -45,35 +45,50 @@ stbt.debug("Arguments:\n" + "\n".join([
...
             'thumbnail.jpg',
             cv2.resize(screenshot, (
                 640, 640 * screenshot.shape[0] // screenshot.shape[1])),
             [cv2.IMWRITE_JPEG_QUALITY, 50])
 
 
+@contextmanager
+def video(args):
+    dut = _stbt.core.new_device_under_test_from_config(args)
+    with stbt._set_dut_singleton(dut), dut:  # pylint: disable=protected-access
```
Comment (drothlis) [inline stbt-run]: Presumably `_set_dut_singleton` is added in a subsequent commit (I'm reviewing these in the order that github presents them which is by commit date not by git dag order, if I recall correctly).

File: _stbt/stbt_run.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env python
```
Comment (drothlis) [inline _stbt/stbt_run.py]: This shouldn't have a #! line.

File: stbt-run [hunk 3]
```diff
@@ -1,21 +1,20 @@
-#!/usr/bin/env python
+#!/usr/bin/python
```
Comment (drothlis) [inline stbt-run]: Why removing `/usr/bin/env`? I think it was made that way explicitly for virtualenv support.

File: _stbt/core.py [hunk 4]
```diff
@@ -1591,6 +1592,9 @@ def argparser():
         default=get_config('global', 'restart_source', type_=bool),
         help='Restart the GStreamer source pipeline when video loss is '
              'detected')
+    parser.add_argument(
+        '--save-video', help='Record video to the specified file',
+        metavar='FILE', default=get_config('run', 'save_video'))
```
[the remark this answers is not in the document]
Comment (wmanley) [reply _stbt/core.py]: Yeah.  I thought about breaking this function up a bit.  We could have "add_videosrc_args", "add_videosink_args", etc, but I'm not sure it's worth it.  As long as the defaults haven't changed I think it's fine.

File: _stbt/state_watch.py [hunk 5]
```diff
@@ -83,6 +83,12 @@ class StateSender(object):
     def __init__(self, file_):
         self._file = file_
 
+    def __enter__(self):
+        pass
+
+    def __exit__(self, _, _1, _2):
+        self.close()
+
```
[the remark this answers is not in the document]
Comment (wmanley) [reply _stbt/state_watch.py]: You're right, this should be in "stbt-run refactor: Factor out tracing"

File: _stbt/stbt_run.py (hunk 2 above)
Comment (wmanley) [reply _stbt/stbt_run.py]: Fixed

File: stbt-run (hunk 3 above)
Comment (wmanley) [reply stbt-run]: A mistake, fixed.

File: tests/test-stbt-py.sh [hunk 6]
```diff
@@ -638,6 +638,7 @@ test_that_transformation_pipeline_transforms_video() {
 
 test_multithreaded() {
     cat > test.py <<-EOF &&
+	import sys
 	import time
```
[the remark this answers is not in the document]
Comment (wmanley) [reply tests/test-stbt-py.sh]: Fixed

[the remark this answers is not in the document]
Comment (wmanley) [reply tests/test-stbt-py.sh]: Fixed

File: stbt-run [hunk 7]
```diff
@@ -114,47 +114,66 @@ def import_by_filename(filename_):
...
+        filename = os.path.abspath(script)
+
+        test_globals = {
+            '__builtins__': __builtins__,
+            '__name__': '__main__',
+            '__file__': script,
+            '__doc__': None,
+            '__package__': None,
+            'stbt': stbt,
+        }
```
[the remark this answers is not in the document]
Comment (wmanley) [reply stbt-run]: Yeah, I was rather pleased with this.  I go the list of default globals with:

```
$ echo print globals() >moo.py
$ python moo.py
{'__builtins__': <module '__builtin__' (built-in)>, '__name__': '__main__', '__file__': 'moo.py', '__doc__': None, '__package__': None}

```

File: _stbt/state_watch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wmanley) [reply _stbt/state_watch.py]: Fixed in rebase

File: _stbt/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (drothlis) [reply _stbt/core.py]: @wmanley says: All of the arguments used by `new_device_under_test_from_config` are set by this `argparser` function. That's why it needs to go here, not in `stbt-run`.

[the remark this answers is not in the document]
Comment (drothlis) [reply _stbt/core.py]: Minor correction: This doesn't affect `stbt-control`. I had grepped for `argparser()` but `stbt-control` has its own implementation of `argparser()`.

**Pull request merged**